Add Controller+Mapper+MockGCP for AnywhereCache resource.

googlyrahman opened the pull request:
### Change description

1. Add AnywhereCache controller + mapper + fuzzer
2. Add mockgcp tests

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_controller.go:
```diff
@@ -0,0 +1,467 @@
+
+	if status.ObservedState.PendingUpdate != nil && *status.ObservedState.PendingUpdate {
+		// Metadata updates are performed via LROs, hence the status is 'Updating' instead of 'UpToDate' immediately.
+		return updateOp.UpdateStatus(ctx, status, getCondition(v1.ConditionTrue, k8s.Updating, k8s.UpdatingMessage))
```
I'm seeking feedback here, should we keep `ConditionTrue`, or `ConditionFalse` when the status is Updating?

googlyrahman replied:
It should be "False", Hence updated the same.

googlyrahman commented:
/assign @acpana 

The first 3 commit is from another PR https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/4573, Hence please focus from the 4th one.

googlyrahman commented:
Failure looks unrelated, ```--- FAIL: TestAllInSeries/fixtures/sensitivevaluesimple (17.14s)```

acpana commented on mockgcp/mockstorage/storagecontrol.go:
```diff
@@ -0,0 +1,197 @@
+
+	now := time.Now()
+	obj.UpdateTime = timestamppb.New(now)
+	obj.State = "paused"
```
are there proto constants that we can leverage here?

googlyrahman replied:
Define constants, and used here.

acpana commented on pkg/test/resourcefixture/testdata/basic/storage/v1alpha1/storageanywherecache/storageanywherecache/_generated_object_storageanywherecache.golden.yaml:
I think we want to create 2 fixture tests here:

-base/ -basic
-full

Not for base we create the object with the minimum amount of fields needed (usually just the required ones). Then update the required fields (at least for those that are updatable).

acpana reviewed (commented):
This is coming along nicely 💯 

Left some small questions. Please also think about including a scenarios test. It doesn't have to be part of this PR 😛

googlyrahman commented on pkg/k8s/constants.go:
```diff
@@ -37,6 +37,8 @@ const (
 	UpToDateMessage                      = "The resource is up to date"
 	Created                              = "Created"
+	Creating                             = "Creating"
+	CreatingMessage                      = "The resource is under creation"
```
Done

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_fuzzer.go:
```diff
@@ -0,0 +1,49 @@
+// Copyright 2024 Google LLC
```
Done

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_mappings.go:
```diff
@@ -0,0 +1,85 @@
+
+		out.BucketRef = &refs.StorageBucketRef{External: parent.String()}
+		out.ResourceID = &resourceID
+	} else {
+		// dead code, here to make fuzzer test happy.
+		// Fuzzer test sends empty values, and we can't throw error.
+		// As throwing the same, would cause fuzzer test to fail.
+	}
```
The fuzzer sets name="" in `pb.AnywhereCache`, whereas we expect the name to be in the format `projects/_/....` Since this expectation is not met, an error is thrown and the fuzzer test fails.

However, in real-world scenarios, this situation would never occur. Therefore, this is effectively dead code

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_controller.go:
```diff
@@ -0,0 +1,471 @@
+	return a.GetCurrentState() != a.GetDesiredState()
+}
+
+// UpdateState changes the cache state to the desired state if different from current.
+func (a *AnywhereCacheAdapter) UpdateState(ctx context.Context) error {
```
Isn't this tested via `mockgcp` tests? 

Since this function consists of gcp calls, it might not be simple to mock it! do we've any example of  mocking gcp call?

googlyrahman replied:
Added scenario tests, which tests this part of the code.

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_controller.go:
```diff
@@ -0,0 +1,471 @@
+}
+
+// getCondition is a helper to create a KRM condition for resource status.
+func getCondition(status v1.ConditionStatus, reason string, message string) *v1alpha1.Condition {
```
Done

googlyrahman commented on config/tests/samples/create/samples.go:
```diff
@@ -193,6 +193,10 @@ func waitForReadySingleResource(t *Harness, u *unstructured.Unstructured, timeou
 		return
 	}
 
+	if u.GetKind() == "StorageAnywhereCache" {
+		timeout = 4 * time.Hour
+	}
```
Added the resource to list specified at `LONG_RUNNING_CRUD_TESTS_REGEX`.

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_controller.go:
```diff
@@ -0,0 +1,471 @@
+
+	// dead code, but kept it here for mockgcp tests.
+	return createOp.UpdateStatus(ctx, status, getCondition(v1.ConditionTrue, k8s.UpToDate, k8s.UpToDateMessage))
+}
```
Just kept it in case if cache creation becomes instantaneous in future, we can throw error too - let me know what should we do here?

acpana replied:
I'd remove this comment! it's confusing and I think we definitely hit the code block once the cache is created in GCP

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_controller.go:
```diff
@@ -0,0 +1,471 @@
+		If a user requests a state change, we prioritize and perform it immediately,
+		as such changes are generally possible regardless of any ongoing metadata updates.
+	*/
+	if a.IsStateChangeRequested() {
```
Added scenario tests! already documented while writing spec help text. See this https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_storageanywherecaches.storage.cnrm.cloud.google.com.yaml#L102

googlyrahman commented:
> This is coming along nicely 💯
> 
> Left some small questions. Please also think about including a scenarios test. It doesn't have to be part of this PR 😛

Added scenario tests, and addressed comments.

googlyrahman commented:
Triggered tests by force push, as failure was unrelated.

acpana commented on scripts/shared-vars-public.sh:
```diff
@@ -111,11 +111,13 @@ SUPPORTED_SERVICES=(
 # Regex used to match long running tests cases (10m+ runtime). Any new
 # long-running tests should be added to this regex in alphabetical order.
-LONG_RUNNING_CRUD_TESTS_REGEX="basicalloydbbackup|\
+LONG_RUNNING_CRUD_TESTS_REGEX="storageanywherecache|\
```
I think this is storageanywherecache-basic AND storageanywherecache-full; can you also alphabetize it?

acpana reviewed (commented):
/lgtm

This overall LGTM, a few nits, please address 🙏🏼 

cc @yuwenma please also review and "approve" if possible -- this is an alpha resource still!

yuwenma reviewed (commented):
A huge progress and great job, @googlyrahman!

I left some comment about declarative API improvements and questions to double check the corner case handling.

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_controller.go:
```diff
@@ -0,0 +1,476 @@
+	}
+
+	name := fmt.Sprintf("%s/anywhereCaches/%s", a.id.Parent().String(), metadata.GetAnywhereCacheId())
+	resource, err := a.gcpClient.GetAnywhereCache(ctx, &pb.GetAnywhereCacheRequest{Name: name})
```
Fortunately, doing a GET returns the entire payload 😛 irrespective of LRO progress - So we should be fine here?

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_controller.go:
```diff
@@ -0,0 +1,476 @@
+	anywhereCacheStateRunning  = "running"
+	anywhereCacheStatePaused   = "paused"
+	anywhereCacheStateDisabled = "disabled"
+	anywhereCacheStateInvalid  = "invalid" // For cases where the state is unexpected
```
The [cloud.google.com/go/storage/control](https://pkg.go.dev/cloud.google.com/go/storage/control/apiv2/controlpb) package doesn't define specific variables for state; instead, the proto stores state as a string, not an enum. See https://screenshot.googleplex.com/A8pE3BvrMbR25VC

let me know if i am missing anything here?

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_controller.go:
```diff
@@ -0,0 +1,476 @@
+		return fmt.Errorf("cannot change state from or to '%s' state", anywhereCacheStateCreating)
+	}
+
+	var err error
```
You can find this particular code here [this](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/b8bf1bc71ace7fb806d3bef3ca5ebfa1f0db5036/pkg/controller/direct/storage/control/anywherecache_controller.go#L423)

If `desiredState == running`, and `actualState == disabled`, it calls `Resume` which enables the cache from disabled back to running state.

Also there should be a scenario test for the same.

googlyrahman commented:
> After looking at the following code, I suggest we limit the desired state to be disable, pause and resume, and also consider separating disabled from pause/resume. Otherwise, the anywhere cache will hit some issues, just to name a few that comes to my mind:

"disable", "pause", and "resume" are operations, and not state, the value of state can be "running", "disabled", "paused", and IMO it is not a good idea to keep an operation value in place of variable which stores state, i.e "desiredState"

Additionally "state" is just like a metadata field, but it can't be updated via normal updateAPI endpoint, instead it has other endpoints.

> Setting "running" as desired and default, will interrupt another KCC operation "unset" and "unmanaged". Once the API is released to Beta, we can hardly change that.

I can't find any "unset" or "unmanaged" operation in kubectl command, there's "set" operation. Am i missing anything here? If yes, would you mind pointing me to documentation of these KCC operation. Also I don't get how this field `desiredState` would interrupt these operations? can you add details to it.

> The definition of "running" state is a duplication of the existence of the anywhereCache K8s object itself and we want to get rid of that.

Not exactly, it's cache runtime behavior whether the cache is running, paused, or disabled. A cache does exists in DISABLED state too, so does that mean we should kick "disable" too?
 
> disable and pause/resume normally have different behavior and expectations on the server side (TTL, termination state, etc), we want to distinguish them from the configurations so that users can set the specification more flexible. For example, if using a single field to do "re-enable a disabled-but-not-deleted object and make sure it paused when enabled" , users have to consider the order of "re-enable", "pause", "resume", (and use the value "pause" because it's the final desired state). This is imperative thinking, but the KCC api should be declarative.

From above comment: `users have to consider the order of "re-enable", "pause", "resume"`

-> Not exactly, this is all managed at controller level - User do not need to think of any order to change the state, This order is handled internally at the code. See https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/b8bf1bc71ace7fb806d3bef3ca5ebfa1f0db5036/pkg/controller/direct/storage/control/anywherecache_controller.go#L428 - You can additionally look at scenario tests, and confirm that ordering is being handled at controller level.  On side note, "disable" -> "running" can be done in one operation, `ResumeAnywhereCacheCall`.

I think that answers that this would be a declarative rather than imperative.

yuwenma reviewed (commented):
/lgtm
/approve

Thanks! I left some suggestions about code readability improvement. Non blocker to the PR. One thing we do want to fix is the chagne in file `[tests/e2e/unified_test.go](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/4593/files/08750dd1859cb812a34f90a11ac51c2f0ca47d45#diff-45e74dc3e22d6fbc3f979edfe4959d40c702bf51bea40854de4e4ab393bb5330)` which can be done in follow up PRs.

googlyrahman commented on pkg/controller/direct/storage/control/anywherecache_controller.go:
```diff
@@ -0,0 +1,476 @@
+		return fmt.Errorf("Failed to GET anywhere cache metadata: %w", err)
+	}
+
+	name := fmt.Sprintf("%s/anywhereCaches/%s", a.id.Parent().String(), metadata.GetAnywhereCacheId())
```
Sure, See - https://screenshot.googleplex.com/AzKUtxbf34QJR2m, and https://screenshot.googleplex.com/8MQhqFeJTjbeJEk

https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/b40454d67d149dbb3733e7a8fbf5d65a44b5069f/pkg/controller/direct/storage/control/anywherecache_controller.go#L215

googlyrahman commented on tests/e2e/unified_test.go:
```diff
@@ -579,6 +579,14 @@ func isOperationDone(s string) bool {
 
 // addTestTimeout will ensure the test fails if not completed before timeout
 func addTestTimeout(ctx context.Context, t *testing.T, timeout time.Duration, name string) context.Context {
+	if name == "storageanywherecache" {
```
Okay @yuwenma, let's go ahead and submit this as is? We want to observe how the e2e tests are invoked in postgrids, especially when mocks are replaced by actual e2e gcp targets. Meanwhile, I can add a presubmit specifically for storage?

does that sounds good?

acpana commented:
/lgtm

let's follow up on Yuwen's comments please 🙏🏼 

Thanks for fixing the fuzzer too 💯 🏅
